osu.Game/Overlays/DirectOverlay.cs

```diff
@@ -225,28 +249,122 @@ protected override void PopIn()
             base.PopIn();
 
             // Queries are allowed to be run only on the first pop-in
-            if (getSetsRequest == null)
+            if (!beatmapSetPager.IsFetching)
                 queueUpdateSearch();
         }
 
-        private SearchBeatmapSetsRequest getSetsRequest;
-
         private readonly Bindable<string> currentQuery = new Bindable<string>(string.Empty);
 
         private ScheduledDelegate queryChangedDebounce;
+        private ScheduledDelegate addPageDebounce;
         private PreviewTrackManager previewTrackManager;
+        private BeatmapSetPager beatmapSetPager;
+
+        private class BeatmapSetPager
+        {
+            public event PageFetchHandler PageFetch;
+
+            private readonly RulesetStore rulesets;
+
+            private SearchBeatmapSetsRequest getSetsRequest;
+
+            private int currentPage = 1;
+
+            public bool IsLastPageFetched { get; private set; } = false;
+            public bool IsFetching => getSetsRequest != null;
+
+            public BeatmapSetPager(RulesetStore rulesets)
+            {
+                this.rulesets = rulesets;
+            }
+
+            public SearchBeatmapSetsRequest FetchNextPage(string query, RulesetInfo ruleset, BeatmapSearchCategory searchCategory = BeatmapSearchCategory.Any, DirectSortCriteria sortCriteria = DirectSortCriteria.Ranked, SortDirection direction = SortDirection.Descending)
```

Comment:
This operation will be wrong if somebody switches the criteria between `FetchNextPage` calls, since it won't reset the page counter. The criteria should probably come in through the ctor (so instances will be single-filter-use-only), or be set with a separate method that will call `Reset()`.

Reply:
The query is now provided in the constructor.